### PR TITLE
Locks around data track

### PR DESCRIPTION
**Pull request by boks1971**
@cnderrauber Setting up this PR to get your :eyes: I am putting locks and making sure the callback happens outside the lock. I also added an ID method to data track interface so that there is no need to cast. But, data track does not work after migration with these changes. Don't know what is causing that. But, wanted to get your thoughts.

### CODE REVIEW

File: pkg/rtc/participant.go [hunk 1]
```diff
@@ -499,9 +501,16 @@ func (p *ParticipantImpl) Close(sendLeave bool) error {
 		})
 	}
 
-	if p.dataTrack != nil {
-		p.dataTrack.Close()
+	var dt *DataTrack
+	p.lock.Lock()
+	dt = p.dataTrack
+	p.dataTrack = nil
```
Comment (boks1971) [inline pkg/rtc/participant.go:507]: previously this was not set to `nil` after closing. And there is no `AddOnClose` registered on the data track also. So, set it to `nil` here.

File: pkg/rtc/participant.go [hunk 2]
```diff
@@ -1085,15 +1094,25 @@ func (p *ParticipantImpl) OnDataTrackPublished(f func(types.LocalParticipant, ty
...
+	}
+}
+
+func (p *ParticipantImpl) onDataChannelLocked(dc *webrtc.DataChannel) (created bool) {
 	if p.State() == livekit.ParticipantInfo_DISCONNECTED {
 		return
 	}
 	if p.dataTrack == nil {
 		p.dataTrack = NewDataTrack(livekit.TrackID("DT_"+p.params.SID), p.params.SID, p.params.Logger)
-		if p.onDataTrackPublished != nil {
```
Comment (boks1971) [inline pkg/rtc/participant.go:1093]: Moving this out to call sites so that this can be called outside the lock.

File: pkg/rtc/participant.go [hunk 3]
```diff
@@ -1618,12 +1639,15 @@ func (p *ParticipantImpl) DebugInfo() map[string]interface{} {
 }
 
 func (p *ParticipantImpl) GetDataTrack() types.DataTrack {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
 	return p.dataTrack
```
Comment (boks1971) [inline pkg/rtc/participant.go:1645]: I am guessing we might need the same interface check here?

Comment (cnderrauber) [reply pkg/rtc/participant.go:1645]: yeah, need same check here

File: pkg/rtc/participant.go [hunk 4]
```diff
@@ -1651,10 +1675,17 @@ func (p *ParticipantImpl) handlePendingDataChannels() {
 		if err != nil {
 			p.params.Logger.Errorw("create migrated data channel failed", err, "label", ci.Label)
 		} else if dc != nil {
-			p.onDataChannel(dc)
+			created = created || p.onDataChannelLocked(dc)
```
Comment (cnderrauber) [inline pkg/rtc/participant.go:1678]: if created is true, this expression will be short circuit, and p.OnDataChannelLocked will not be execute.
we have two dc here(lossy and reliable), so p.onDataChannelLocked will be only execute once, and another dc's OnMessage will not be listened for datatrack

**Comment by cnderrauber** [comment]
Looks good! left a suggestion about migration

File: pkg/rtc/participant.go (hunk 4 above)
Comment (boks1971) [reply pkg/rtc/participant.go]: Ah, missed that. Fixed it in this commit - https://github.com/livekit/livekit-server/pull/508/commits/ef1cdcf58eab0055ba4c17553c838624eb96f2cf. Migration works now. Thank you @cnderrauber

File: pkg/rtc/participant.go (hunk 3 above)
Comment (boks1971) [reply pkg/rtc/participant.go]: made the change in this commit - https://github.com/livekit/livekit-server/pull/508/commits/ef1cdcf58eab0055ba4c17553c838624eb96f2cf

**Review by cnderrauber — APPROVED** [review]

**Pull request merged**